### PR TITLE
SAI/Integrated Standy Instrument Self Test

### PR DESCRIPTION
**Pull request by 2hwk**
None

**Summary of Changes**
* Added Self Test for SAI / Integrated Standby Instrument System - On timer of 73-75 Seconds. Activates when aircraft is powered on (batt or APU).
* Is already completed when starting from takeoff on runway

**Screenshots (if necessary)**
Before:
image 

After:
image

See: https://youtu.be/g_Mq0mI50GM

**Additional context**
https://youtu.be/g_Mq0mI50GM

Self-Test functionality has been modeled upon this reference video. 

Timer is given a slight random value for flavor, and is reset when all power is lost. 

This may not be 100% realistic behavior yet. 

Fonts are definitely not correct, will need more investigation to get this right.

### CODE REVIEW

File: A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/SAI/A320_Neo_SAI.css [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+@font-face {
+  font-family: "Roboto-Bold";
+  src: url("/Fonts/RobotoMono-Bold.ttf") format("truetype");
+  font-weight: normal;
+  font-style: normal; }
+
+a320-neo-sai {
+  width: 100%;
+  height: 100%;
+  overflow: hidden; }
```
Comment (Meqolo) [inline A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/SAI/A320_Neo_SAI.css]: Closing bracket should be on next line

File: A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/SAI/A320_Neo_SAI.js [hunk 2]
```diff
@@ -1002,5 +1003,209 @@ class A320_Neo_SAI_PressureIndicator extends HTMLElement {
...
+    onEnter() {
+    }
+    isReady() {
+        return true;
+    }
+    onUpdate(_deltaTime) {
+
+        // Delta time mitigation strategy
+        let curTime = Date.now();
+        let localDeltaTime = curTime - this._lastTime;
```
Comment (Benjozork) [inline A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/SAI/A320_Neo_SAI.js:1021]: Both can be `const`.

File: A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/SAI/A320_Neo_SAI.js [hunk 3]
```diff
@@ -1002,5 +1003,209 @@ class A320_Neo_SAI_PressureIndicator extends HTMLElement {
...
+        this.testTimer = Math.floor(Math.random() * 3) + this.start_time;
+        this.complete = false;
+        let element = document.querySelector("#SelfTestHider");
+        element.style.display = "none";
+        this.selfTestDiv.style.display = "block";
+    }
+    finishTest() {
+        this.testTimer = 0;
+        this.selfTestDiv.style.display = "none";
+        let element = document.querySelector("#SelfTestHider");
```
Comment (Benjozork) [inline A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/SAI/A320_Neo_SAI.js:1202]: Can be `const`.

```suggestion
        const element = document.querySelector("#SelfTestHider");
```

File: A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/SAI/A320_Neo_SAI.js [hunk 4]
```diff
@@ -1002,5 +1003,209 @@ class A320_Neo_SAI_PressureIndicator extends HTMLElement {
...
+            if (this.testTimer <= 0) {
+                this.finishTest();
+            }
+        }
+    }
+
+    resetTimer() {
+        this.testTimer = Math.floor(Math.random() * 3) + this.start_time;
+        this.complete = false;
+        let element = document.querySelector("#SelfTestHider");
```
Comment (Benjozork) [inline A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/SAI/A320_Neo_SAI.js:1195]: Can be `const`.

```suggestion
        const element = document.querySelector("#SelfTestHider");
```

File: A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/SAI/A320_Neo_SAI.js [hunk 5]
```diff
@@ -1002,5 +1003,209 @@ class A320_Neo_SAI_PressureIndicator extends HTMLElement {
...
+    update(dTime) {
+        if (!this.complete) {
+            if (this.testTimer >= 0) {
+                this.testTimer -= dTime / 1000;
+            }
+
+            if (this.testTimer > 9) {
+                this.st_tmr_txt.textContent = "INIT " + Math.ceil(this.testTimer) + "s";
+            }
+            else {
```
Comment (Benjozork) [inline A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/SAI/A320_Neo_SAI.js:1183]: `else` should be on the same line.

File: A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/SAI/A320_Neo_SAI.js [hunk 6]
```diff
@@ -1002,5 +1003,209 @@ class A320_Neo_SAI_PressureIndicator extends HTMLElement {
...
+        let st_att = document.createElementNS(Avionics.SVG.NS, "rect");
+        st_att.setAttribute("id", "AttTest")
+        st_att.setAttribute("fill", "#afbb3a");
+        st_att.setAttribute("x", "36%");
+        st_att.setAttribute("y", boxRow1 + "%");
+        st_att.setAttribute("width", boxWidth + "%");
+        st_att.setAttribute("height", boxHeight + "%");
+        this.selfTestSVG.appendChild(st_att);
+
+        let st_att_txt = document.createElementNS(Avionics.SVG.NS, "text");
```
Comment (Benjozork) [inline A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/SAI/A320_Neo_SAI.js:1162]: Can be `const`.

```suggestion
        const st_att_txt = document.createElementNS(Avionics.SVG.NS, "text");
```

File: A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/SAI/A320_Neo_SAI.js [hunk 7]
```diff
@@ -1002,5 +1003,209 @@ class A320_Neo_SAI_PressureIndicator extends HTMLElement {
...
+        this.st_tmr_txt.textContent = "INIT "+ Math.ceil(this.testTimer)  +"s";
+        this.st_tmr_txt.setAttribute("font", "Roboto");
+        this.st_tmr_txt.setAttribute("font-weight", "900");
+        this.st_tmr_txt.setAttribute("font-size", "42px");
+        this.st_tmr_txt.setAttribute("fill", "#1f242d");
+        this.st_tmr_txt.setAttribute("x", "31%");
+        this.st_tmr_txt.setAttribute("y", boxRow3 + txt_off + "%");
+        this.selfTestSVG.appendChild(this.st_tmr_txt);
+
+        let st_att = document.createElementNS(Avionics.SVG.NS, "rect");
```
Comment (Benjozork) [inline A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/SAI/A320_Neo_SAI.js:1153]: Can be `const`.

```suggestion
        const st_att = document.createElementNS(Avionics.SVG.NS, "rect");
```

File: A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/SAI/A320_Neo_SAI.js [hunk 8]
```diff
@@ -1002,5 +1003,209 @@ class A320_Neo_SAI_PressureIndicator extends HTMLElement {
...
+        st_alt_txt.textContent = "ALT";
+        st_alt_txt.setAttribute("font", "Roboto");
+        st_alt_txt.setAttribute("font-weight", "900");
+        st_alt_txt.setAttribute("font-size", "42px");
+        st_alt_txt.setAttribute("fill", "#1f242d");
+        st_alt_txt.setAttribute("x", "72.5%");
+        st_alt_txt.setAttribute("y", boxRow2 + txt_off + "%");
+        this.selfTestSVG.appendChild(st_alt_txt);
+
+        let st_tmr = document.createElementNS(Avionics.SVG.NS, "rect");
```
Comment (Benjozork) [inline A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/SAI/A320_Neo_SAI.js:1133]: Can be `const`.

```suggestion
        const st_tmr = document.createElementNS(Avionics.SVG.NS, "rect");
```

File: A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/SAI/A320_Neo_SAI.js [hunk 9]
```diff
@@ -1002,5 +1003,209 @@ class A320_Neo_SAI_PressureIndicator extends HTMLElement {
...
+        let st_alt = document.createElementNS(Avionics.SVG.NS, "rect");
+        st_alt.setAttribute("id", "AltTest");
+        st_alt.setAttribute("fill", "#afbb3a");
+        st_alt.setAttribute("x", "70%");
+        st_alt.setAttribute("y", boxRow2 + "%");
+        st_alt.setAttribute("width", boxWidth + "%");
+        st_alt.setAttribute("height", boxHeight + "%");
+        this.selfTestSVG.appendChild(st_alt);
+
+        let st_alt_txt = document.createElementNS(Avionics.SVG.NS, "text");
```
Comment (Benjozork) [inline A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/SAI/A320_Neo_SAI.js:1122]: Can be `const`.

```suggestion
        const st_alt_txt = document.createElementNS(Avionics.SVG.NS, "text");
```

File: A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/SAI/A320_Neo_SAI.js [hunk 10]
```diff
@@ -1002,5 +1003,209 @@ class A320_Neo_SAI_PressureIndicator extends HTMLElement {
...
+        st_spd_txt.textContent = "SPD";
+        st_spd_txt.setAttribute("font", "Roboto");
+        st_spd_txt.setAttribute("font-weight", "900");
+        st_spd_txt.setAttribute("font-size", "42px");
+        st_spd_txt.setAttribute("fill", "#1f242d");
+        st_spd_txt.setAttribute("x", "8.75%");
+        st_spd_txt.setAttribute("y", boxRow2 + txt_off + "%");
+        this.selfTestSVG.appendChild(st_spd_txt);
+
+        let st_alt = document.createElementNS(Avionics.SVG.NS, "rect");
```
Comment (Benjozork) [inline A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/SAI/A320_Neo_SAI.js:1113]: Can be `const`.

```suggestion
        const st_alt = document.createElementNS(Avionics.SVG.NS, "rect");
```

File: A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/SAI/A320_Neo_SAI.js [hunk 11]
```diff
@@ -1002,5 +1003,209 @@ class A320_Neo_SAI_PressureIndicator extends HTMLElement {
...
+        let st_spd = document.createElementNS(Avionics.SVG.NS, "rect");
+        st_spd.setAttribute("id", "SpeedTest");
+        st_spd.setAttribute("fill", "#afbb3a");
+        st_spd.setAttribute("x", "8%");
+        st_spd.setAttribute("y", boxRow2 + "%");
+        st_spd.setAttribute("width", boxWidthSmall + "%");
+        st_spd.setAttribute("height", boxHeight + "%");
+        this.selfTestSVG.appendChild(st_spd);
+
+        let st_spd_txt = document.createElementNS(Avionics.SVG.NS, "text");
```
Comment (Benjozork) [inline A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/SAI/A320_Neo_SAI.js:1102]: Can be `const`.

```suggestion
        const st_spd_txt = document.createElementNS(Avionics.SVG.NS, "text");
```

File: A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/SAI/A320_Neo_SAI.css (hunk 1 above)
Comment (AdenFlorian) [reply A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/SAI/A320_Neo_SAI.css]: asobo convention is to have it on same line as last property, so this is fine

File: A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/SAI/A320_Neo_SAI.js [hunk 12]
```diff
@@ -1002,5 +1003,209 @@ class A320_Neo_SAI_PressureIndicator extends HTMLElement {
...
+        this.selfTestSVG.setAttribute("id", "SelfTestSVG");
+        this.selfTestSVG.setAttribute("viewBox", "0 0 600 600");
+        this.selfTestSVG.style.position = "absolute";
+        this.selfTestSVG.style.top = "0%";
+        this.selfTestSVG.style.left = "0%";
+        this.selfTestSVG.style.width = "100%";
+        this.selfTestSVG.style.height = "100%";
+        this.selfTestDiv.appendChild(this.selfTestSVG);
+
+        let st_spd = document.createElementNS(Avionics.SVG.NS, "rect");
```
Comment (Benjozork) [inline A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/SAI/A320_Neo_SAI.js:1093]: Can be `const`.

```suggestion
        const st_spd = document.createElementNS(Avionics.SVG.NS, "rect");
```

File: A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/SAI/A320_Neo_SAI.js [hunk 13]
```diff
@@ -1002,5 +1003,209 @@ class A320_Neo_SAI_PressureIndicator extends HTMLElement {
...
+        const boxWidthInit = 30;
+        const boxRow1 = 32.5;
+        const boxRow2 = 46.5;
+        const boxRow3 = 64;
+        const txt_off = 6;
+
+        this.testTimer = Math.floor(Math.random() * 3) + this.start_time;
+        Utils.RemoveAllChildren(this);
+
+        let element = document.querySelector("#SelfTestHider");
```
Comment (Benjozork) [inline A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/SAI/A320_Neo_SAI.js:1069]: Can be `const`.

```suggestion
        const element = document.querySelector("#SelfTestHider");
```

Also, try finding a more descriptive name ;)

File: A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/SAI/A320_Neo_SAI.js [hunk 14]
```diff
@@ -1002,5 +1003,209 @@ class A320_Neo_SAI_PressureIndicator extends HTMLElement {
...
+    onEvent(_event) {
+    }
+}
+class A320_Neo_SAI_SelfTestTimer extends HTMLElement {
+    connectedCallback() {
+        this.construct();
+    }
+
+    construct() {
+
```
Comment (Benjozork) [inline A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/SAI/A320_Neo_SAI.js:1054]:
```suggestion
```

File: A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/SAI/A320_Neo_SAI.js [hunk 15]
```diff
@@ -1002,5 +1003,209 @@ class A320_Neo_SAI_PressureIndicator extends HTMLElement {
...
+            this.selfTestElement.finishTest();
+        }
+
+    }
+    onExit() {
+    }
+    onEvent(_event) {
+    }
+}
+class A320_Neo_SAI_SelfTestTimer extends HTMLElement {
```
Comment (Benjozork) [inline A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/SAI/A320_Neo_SAI.js:1048]:
```suggestion
class A320_Neo_SAI_SelfTestTimer extends HTMLElement {

```

File: A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/SAI/A320_Neo_SAI.js [hunk 16]
```diff
@@ -1002,5 +1003,209 @@ class A320_Neo_SAI_PressureIndicator extends HTMLElement {
...
+        let curTime = Date.now();
+        let localDeltaTime = curTime - this._lastTime;
+        this._lastTime = curTime;
+
+        const ac_pwr = SimVar.GetSimVarValue("L:ACPowerAvailable", "bool");
+        const dc_pwr = SimVar.GetSimVarValue("L:DCPowerAvailable", "bool");
+        const ColdAndDark = SimVar.GetSimVarValue('L:A32NX_COLD_AND_DARK_SPAWN', 'Bool');
+
+        if ( (ac_pwr || dc_pwr) && !this.complete) {
+            //this.selfTestElement.update(_deltaTime);
```
Comment (Benjozork) [inline A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/SAI/A320_Neo_SAI.js:1029]: Remove commented code ;)

File: A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/SAI/A320_Neo_SAI.js [hunk 17]
```diff
@@ -1002,5 +1003,209 @@ class A320_Neo_SAI_PressureIndicator extends HTMLElement {
...
+        // Delta time mitigation strategy
+        let curTime = Date.now();
+        let localDeltaTime = curTime - this._lastTime;
+        this._lastTime = curTime;
+
+        const ac_pwr = SimVar.GetSimVarValue("L:ACPowerAvailable", "bool");
+        const dc_pwr = SimVar.GetSimVarValue("L:DCPowerAvailable", "bool");
+        const ColdAndDark = SimVar.GetSimVarValue('L:A32NX_COLD_AND_DARK_SPAWN', 'Bool');
+
+        if ( (ac_pwr || dc_pwr) && !this.complete) {
```
Comment (Benjozork) [inline A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/SAI/A320_Neo_SAI.js:1028]:
```suggestion
        if ((ac_pwr || dc_pwr) && !this.complete) {
```

File: A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/SAI/A320_Neo_SAI.js [hunk 18]
```diff
@@ -1002,5 +1003,209 @@ class A320_Neo_SAI_PressureIndicator extends HTMLElement {
...
+
+    resetTimer() {
+        this.testTimer = Math.floor(Math.random() * 3) + this.start_time;
+        this.complete = false;
+        let element = document.querySelector("#SelfTestHider");
+        element.style.display = "none";
+        this.selfTestDiv.style.display = "block";
+    }
+    finishTest() {
+        this.testTimer = 0;
```
Comment (AdenFlorian) [inline A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/SAI/A320_Neo_SAI.js:1200]: can you add a `if (this.complete) return;` here?

File: A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/SAI/A320_Neo_SAI.js (hunk 16 above)
Comment (AdenFlorian) [inline A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/SAI/A320_Neo_SAI.js:1029]: you can delete the comment

File: A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/SAI/A320_Neo_SAI.js [hunk 19]
```diff
@@ -1002,5 +1003,209 @@ class A320_Neo_SAI_PressureIndicator extends HTMLElement {
 }
 customElements.define('a320-neo-sai-pressure-indicator', A320_Neo_SAI_PressureIndicator);
 
+class A320_Neo_SAI_SelfTest extends NavSystemElement {
+    init(root) {
+        this._lastTime = Date.now();
+        this.selfTestElement = this.gps.getChildById("SelfTest");
+        this.complete = false;
```
Comment (AdenFlorian) [inline A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/SAI/A320_Neo_SAI.js:1010]: where is this getting set to true?
i see this.complete = true down below but it's in a different class and wouldn't touch this one, right?

File: A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/SAI/A320_Neo_SAI.js [hunk 20]
```diff
@@ -1002,5 +1003,209 @@ class A320_Neo_SAI_PressureIndicator extends HTMLElement {
...
+        }
+    }
+
+    resetTimer() {
+        this.testTimer = Math.floor(Math.random() * 3) + this.start_time;
+        this.complete = false;
+        let element = document.querySelector("#SelfTestHider");
+        element.style.display = "none";
+        this.selfTestDiv.style.display = "block";
+    }
```
Comment (AdenFlorian) [inline A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/SAI/A320_Neo_SAI.js:1198]: can you make it so this doesn't touch the DOM every frame while no power? maybe have 3 states, OFF, TESTING, ON, and don't run this code if it's OFF, and set it to OFF at bottom of this function?

**Review by AdenFlorian — CHANGES_REQUESTED** [review]
left comments about changes

File: A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/SAI/A320_Neo_SAI.js (hunk 5 above)
Comment (AdenFlorian) [reply A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/SAI/A320_Neo_SAI.js:1183]: we dont have a standard for this, as long as its consistent with the code around it

**Comment by AdenFlorian** [comment]
Looks great! We'll test it once the code looks ready!

File: A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/SAI/A320_Neo_SAI.js (hunk 20 above)
Comment (2hwk) [reply A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/SAI/A320_Neo_SAI.js]: What if there's a greater than comparison that checks if the current time (testTimer) is greater than the default starting time (start_time), that skips the logic if true. Would this be acceptable?

File: A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/SAI/A320_Neo_SAI.js (hunk 19 above)
Comment (2hwk) [reply A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/SAI/A320_Neo_SAI.js]: Just a fragment of code from an earlier version where this flag was set in the extended NavSystemElements class, should have been removed

File: A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/SAI/A320_Neo_SAI.js (hunk 18 above)
Comment (2hwk) [reply A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/SAI/A320_Neo_SAI.js:1199]: Done

File: A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/SAI/A320_Neo_SAI.js (hunk 16 above)
Comment (2hwk) [reply A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/SAI/A320_Neo_SAI.js]: Done

Comment (2hwk) [reply A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/SAI/A320_Neo_SAI.js]: Roger

File: A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/SAI/A320_Neo_SAI.js (hunk 20 above)
Comment (2hwk) [reply A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/SAI/A320_Neo_SAI.js]: This is the logic that is in the latest commit.

File: A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/SAI/A320_Neo_SAI.js (hunk 13 above)
Comment (2hwk) [reply A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/SAI/A320_Neo_SAI.js]: Done

File: A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/SAI/A320_Neo_SAI.js (hunk 2 above)
Comment (2hwk) [reply A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/SAI/A320_Neo_SAI.js]: Fixed

**Review by AdenFlorian — APPROVED** [review]
code looks good, ready to test!

**Comment by Coralec** [comment]
Tested and working properly.

**Pull request merged**